### PR TITLE
Implement Blink v3

### PR DESCRIPTION
**Pull request by kvipka**
- -[x] blink in abyss
- -[x] blink by path ( rechecking each step )
- -[x] where player can walk - blink will available there
- -[x] implemented check on dynamic objects (locked doors and etc)
- -[x] implemented total path of blink.

### CODE REVIEW

**Comment by tkrokli** [comment]
If you wanted your text to look different in the commit message, have a look at these helpful pages:
https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#lists
https://help.github.com/articles/markdown-basics/

File: src/server/game/Spells/Spell.cpp [hunk 1]
```diff
@@ -1283,6 +1283,207 @@ void Spell::SelectImplicitCasterDestTargets(SpellEffIndex effIndex, SpellImplici
...
+                    Position pos;
+                    pos.Relocate(destx, desty, destz, orientation);
+                    m_caster->GetFirstCollisionPosition(m_caster->GetDistance(pos.GetPositionX(), pos.GetPositionY(), pos.GetPositionZ()), 0.0f);
+                    dest = SpellDestination(pos.GetPositionX(), pos.GetPositionY(), pos.GetPositionZ(), pos.GetOrientation());
+                    //float range = sqrt((desty - y)*(desty - y) + (destx - x)*(destx - x));
+                    //TC_LOG_ERROR("server", "Blink number 1, from falling, distance of blink = %f", range);
+                }
+                else
+                {
+                    /*//recalculate, we wanna fly above abyss and don't wanna stop to edge =) (if nessesary enable "stop to edge" - comment this lines
```
Comment (DDuarte) [inline src/server/game/Spells/Spell.cpp]: Remove all the commented out code if it's not necessary. This code is simply making things more confusing

**Comment by Keader** [comment]
Need cleaning much unused code.
Tested and working here, congratulations

**Comment by kvipka** [comment]
Unused code removed. Others lines with LOG needed for investigation of possible trouble. Don't touch it.

**Comment by tkrokli** [comment]
To make this PR show the changes from all your 3 commits, you need to squash your commits:
`git rebase -i HEAD~3` (or look up the squash command for your tool, if you use something else)

File: src/server/game/Spells/Spell.cpp [hunk 2]
```diff
@@ -1283,6 +1283,181 @@ void Spell::SelectImplicitCasterDestTargets(SpellEffIndex effIndex, SpellImplici
...
+                        tstY = y + (float(j)*DELTA_Y);
+
+                        if (j < 2)
+                        {
+                            prevZ = z;
+                            newdistance = 0.0f;
+                            totalpath = 0.0f;
+                        }
+                        else
+                        {
```
Comment (MitchesD) [inline src/server/game/Spells/Spell.cpp]: redundant brackets

File: src/server/game/Spells/Spell.cpp [hunk 3]
```diff
@@ -1283,6 +1283,181 @@ void Spell::SelectImplicitCasterDestTargets(SpellEffIndex effIndex, SpellImplici
...
+                            else if (srange3 < srange2)
+                                tstZ = tstZ3 + 0.5f;
+                            else
+                                tstZ = tstZ2 + 0.5f;
+                        }
+
+                        destz = tstZ;
+                        srange = sqrt((tstY - prevY)*(tstY - prevY) + (tstX - prevX)*(tstX - prevX) + (tstZ - prevZ)*(tstZ - prevZ));
+                        totalpath += srange;
+                        //TC_LOG_ERROR("server", "Blink cycle checking coordinates, number of cycle = %i , distance of step = %f, total path = %f", j, srange, totalpath);
```
Comment (MitchesD) [inline src/server/game/Spells/Spell.cpp]: change to TC_LOG_DEBUG and remove comment

Comment (kvipka) [inline src/server/game/Spells/Spell.cpp]: Nope. If user will use DEBUG log in config by standart - server will bring lags each blink. Only by error log config available without lag- it first. And second - this needed only for investigation of blink in specific situations.

Comment (DDuarte) [inline src/server/game/Spells/Spell.cpp]: If it's not needed, remove the commented out code.
You are saying that a single TC_LOG lags your server?

**Comment by killradio** [comment]
@kvipka `if (!m_caster->HasUnitMovementFlag(MOVEMENTFLAG_FALLING) || (z - ground < 25.0f))` you mean all this calculations will apply to each blink's cast? Or blink adds this movement flag, can someone explain?

File: src/server/game/Spells/Spell.cpp (hunk 3 above)
Comment (killradio) [inline src/server/game/Spells/Spell.cpp]: @DDuarte a single TC_LOG inside loop with the worst case of 10 cycles for each cast.

**Comment by tkrokli** [comment]
Are you asking about only the single line you are quoting, or the whole section covered by that if - test?

**Comment by killradio** [comment]
@tkrokli I'm asking about whole pull request, since author didn't provide proper explanation for these calculations.

**Comment by kvipka** [comment]
someone want to delete all commented lines (include functions for debugging possible troubles and reducing the load on server), another wanna hear explanation of each line. If you can not understand their own, for what others questions?

**Comment by killradio** [comment]
No one asked you about explanation every line, explain magic numbers (like `z - ground < 25.0f`, `z - ground > 3.0f`, `float maxtravelDistZ = 2.65f`) and make proper comments. It's just good coding practice. If you won't - hf&gl.

**Comment by kvipka** [comment]
yes

**Comment by Kittnz** [comment]
@kvipka any update on this?

**Comment by Keader** [comment]
Sometimes when using blink, ends falling in the "infinite".

**Comment by Demonid** [comment]
Glyph of Blind seems to not work after this code. A Mage with and without the glyph will blink the exact same distance.

**Comment by kvipka** [comment]
Demonid:
float distance = m_spellInfo->Effects[effIndex].CalcRadius();

In both version (standart and this) in TC distance calculated by this func. This func taking all modifiers. Need investigate this

**Comment by Keader** [comment]
The stairs before Professor Putricide  (ICC) happens enough. In Battlegrounds (Arathi Basin) near the Lumber is also occurring quite

**Comment by kvipka** [comment]
okay, will search it, but will good, if you can find screenshot from players

**Comment by kvipka** [comment]
Demonid, about Glyph:
https://github.com/healthstone/RE/commit/8a4e14bd52f30da183130bb1e2ac408815373f62

**Comment by Demonid** [comment]
@kvipka Dah xD

Good job ;)

**Comment by kvipka** [comment]
Hi again guys!

Founded mistakes in algorithm and fix it : 
https://www.youtube.com/watch?v=WQxAdwt_lvM&feature=youtu.be

Commit will coming soon

**Comment by tkrokli** [comment]
Your video looks good, I hope your code is just as stable and agreeable for the TC developers. :+1:
(-- wish the NPCs could be as good at not falling under the maps as you were in the video `*sigh*`)

**Comment by kvipka** [comment]
https://github.com/healthstone/RE/commit/4ead62bedc029ace7b49a0a05845bc7bb3bf5694

i think, i need to close this PR and prepare new (because so much changes)

File: src/server/game/Spells/Spell.cpp (hunk 3 above)
Comment (Nyeriah) [inline src/server/game/Spells/Spell.cpp]: If the TC_LOG lags it I wonder how's the overall performance impact of this PR

**Pull request closed without merging**